### PR TITLE
?eval and ?play commands

### LINKED ISSUE #15: Create Playground command
Simply, the command should take everything in a codeblock and use the hosted Rust Playground to run it.

Help on the command should explain the structure of the message

#### Existing Impl
https://github.com/ivandardi/RustbotPython/blob/master/bot/cogs/playground.py

### PR DESCRIPTION
**Pull request by tinaun**
i've tried to make these as close as possible to the ?eval and ?play commands available in the other rustbot. ~~the one thing that doesn't work yet is responding with a playground link for big outputs - i hit the same endpoint that the other bot does but it just returns me an empty string. is it whitelisted?~~ nvm i was just sending a get request instead of a post, works now

closes #15

### CODE REVIEW

**Comment by tinaun** [comment]
i also added a new config environment variable so this might break ci

File: src/playground.rs [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+    stdout: String,
+    stderr: String,
+}
+
+fn run_code(args: &Args, code: &str) -> Result<String> {
+    let mut errs = String::new();
+
+    let channel = args.params.get("channel").unwrap_or(&"nightly");
+    let mode = args.params.get("mode").unwrap_or(&"debug");
+    let edition = args.params.get("edition").unwrap_or(&"2018");
```
Comment (technetos) [inline src/playground.rs:140]: Can we use `unwrap_or_else` here instead? No point in doing extra evaluation if its not needed.

File: src/playground.rs [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+    match Mode::from_str(mode) {
+        Ok(m) => request.mode = m,
+        Err(e) => errs += &format!("{}\n", e),
+    }
+
+    match Edition::from_str(edition) {
+        Ok(e) => request.edition = e,
+        Err(e) => errs += &format!("{}\n", e),
+    }
+
```
Comment (technetos) [inline src/playground.rs:158]: If `errors.len() > 0` we can return here before we make the network request.

File: src/playground.rs [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+#[derive(Debug, Deserialize)]
+struct PlayResult {
+    success: bool,
+    stdout: String,
+    stderr: String,
+}
+
+fn run_code(args: &Args, code: &str) -> Result<String> {
+    let mut errs = String::new();
```
Comment (technetos) [inline src/playground.rs:136]: I prefer `errors` over `errs`

File: src/playground.rs [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        .header(header::REFERER, "https://discord.gg/rust-lang")
+        .json(&payload)
+        .send()?;
+
+    let resp: HashMap<String, String> = resp.json()?;
+    debug!("gist response: {:?}", resp);
+
+    resp.get("id")
+        .map(|id| request.url_from_gist(id))
+        .ok_or("no gist found".into())
```
Comment (technetos) [inline src/playground.rs:206]: Can we use `.ok_or_else` here?

File: src/playground.rs [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+use reqwest::header;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::str::FromStr;
+
+#[derive(Debug, Serialize)]
+struct PlaygroundCode {
+    channel: Channel,
+    edition: Edition,
+    code: String,
```
Comment (technetos) [inline src/playground.rs:17]: Can you pass a `&'p str` here instead and make `PlaygroundCode` take a lifetime `PlaygroundCode<'p>`

File: src/commands.rs [hunk 6]
```diff
@@ -33,6 +33,23 @@ impl Commands {
...
+        handler: impl Fn(Args) -> Result<()> + Send + Sync + 'static,
+    ) {
+        self.__add(command, handler, true)
+    }
+
+    fn __add(
+        &mut self,
+        command: &'static str,
+        handler: impl Fn(Args) -> Result<()> + Send + Sync + 'static,
+        hidden: bool,
```
Comment (technetos) [inline src/commands.rs:52]: Whats the motivation here?

**Review by technetos — CHANGES_REQUESTED** [review]

File: src/playground.rs (hunk 2 above)
Comment (tinaun) [reply src/playground.rs:158]: Should these be fatal errors? I didn't want to immediately return if we can still use the input

File: src/commands.rs (hunk 6 above)
Comment (tinaun) [reply src/commands.rs:52]: I felt like i was adding lots of commands that were error cases / variants on the same command and didn't want to clutter up the help menu

File: src/playground.rs (hunk 2 above)
Comment (technetos) [reply src/playground.rs:158]: Oh I hadnt thought of that.  Ok this is fine

File: src/playground.rs [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+
+    let resp = args
+        .http
+        .post("https://play.rust-lang.org/meta/gist/")
+        .header(header::REFERER, "https://discord.gg/rust-lang")
+        .json(&payload)
+        .send()?;
+
+    let resp: HashMap<String, String> = resp.json()?;
+    debug!("gist response: {:?}", resp);
```
Comment (technetos) [inline src/playground.rs:202]: Will this only log if we compile as `debug`?

File: src/playground.rs [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+
+    api::send_reply(&args, message)?;
+    Ok(())
+}
+
+pub fn eval(args: Args) -> Result<()> {
+    let code = args
+        .params
+        .get("code")
+        .ok_or("Unable to retrieve param: query")?;
```
Comment (technetos) [inline src/playground.rs:248]: I think we can use `ok_or_else` here as well

File: src/playground.rs [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+    resp.get("id")
+        .map(|id| request.url_from_gist(id))
+        .ok_or_else(|| "no gist found".into())
+}
+
+pub fn run(args: Args) -> Result<()> {
+    let code = args
+        .params
+        .get("code")
+        .ok_or("Unable to retrieve param: query")?;
```
Comment (technetos) [inline src/playground.rs:213]: I think we can use ok_or_else here as well

**Review by technetos — CHANGES_REQUESTED** [review]

File: src/commands.rs (hunk 6 above)
Comment (technetos) [reply src/commands.rs:52]: Im ok with this for now.  Its kinda hacky and I dont really like it but its a good sign that we need to do better when generating the help menu.

File: src/playground.rs (hunk 8 above)
Comment (tinaun) [reply src/playground.rs:248]: its just a str literal, i don't think it makes any difference for things that aren't function calls

File: src/playground.rs (hunk 7 above)
Comment (tinaun) [reply src/playground.rs:202]: no it depends on the RUST_LOG environment variable

Comment (technetos) [reply src/playground.rs:202]: Ah we use info level in production

File: src/playground.rs [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+    let result = run_code(&args, code)?;
+    api::send_reply(&args, &result)?;
+    Ok(())
+}
+
+pub fn help(args: Args, name: &str) -> Result<()> {
+    let message = format!(
+        "Compile and run rust code. All code is executed on https://play.rust-lang.org.
+```?{} mode={{}} channel={{}} edition={{}} `code` ```
```
[the remark this answers is not in the document]
Comment (technetos) [reply src/playground.rs:223]: Also, I think \`code\` is incorrect.  It should be 
\`\`\`
code
\`\`\`

To accomplish this inside a codeblock (\`\`\` these the things \`\`\`) I think you will need to put zero width spaces between the backticks inside the codeblocks

**Review by technetos — CHANGES_REQUESTED** [review]

**Review by technetos — APPROVED** [review]

**Pull request merged**